x-pack/metricbeat/module/mssql/fetcher.go

```diff
@@ -0,0 +1,145 @@
+}
+
+// Fetcher will make queries concurrently to the database to fetch results. It
+// must be created by each metricset and fed with the queries that must be
+// executed, a SQL implementor and a pointer to the schema to "apply". A
+// metricset could, potentially, need more than one query to fill all its
+// results.
+type Fetcher struct {
+	schema *s.Schema // Schema is the metricset schema to apply to the fetched data to turn it into a result mapstr.
+	db     *sql.DB   // Database to execute queries against.
+	log    *logp.Logger
+
+	queries []string // List of queries to execute concurrently.
+}
+
+// Close closes db connection to the server
+func (f *Fetcher) Close() error {
+	return f.db.Close()
+}
+
+// Report receives a mb.ReporterV2 to send the data to the outputs. Because the operations are common between metricsets
+// it can be extracted to a common function
+func (f *Fetcher) Report(reporter mb.ReporterV2) {
+	var err error
+	for _, q := range f.queries {
+		if err = f.getEventsWithQuery(f.db, q, reporter); err != nil {
+			logp.Error(err)
+		}
+	}
+}
+
+type rowsResult struct {
+	dest        []interface{}
+	rows        *sql.Rows
+	reporter    mb.ReporterV2
+	columnNames []string
+}
+
+// getEventsWithQuery performs the query on the database and converts the rows
+// to an slice of common.MapStr.
+func (f *Fetcher) getEventsWithQuery(db *sql.DB, query string, reporter mb.ReporterV2) error {
+	// Returns the global status, also for versions previous 5.0.2
+	rows, err := db.Query(query)
+	if err != nil {
+		return errors.Wrapf(err, "error performing db query='%v'", query)
+	}
+	defer func() {
+		if err = rows.Close(); err != nil {
+			logp.Error(errors.Wrap(err, "error closing query row"))
+		}
+	}()
+
+	columnNames, err := rows.Columns()
+	if err != nil {
+		return errors.Wrap(err, "error getting column names")
+	}
+
+	rowsResult := &rowsResult{
+		reporter:    reporter,
+		dest:        make([]interface{}, len(columnNames)),
+		rows:        rows,
+		columnNames: columnNames,
+	}
+
+	for rowsResult.rows.Next() {
+		if err = f.handleRowResult(rowsResult); err != nil {
+			return errors.Wrap(err, "could not convert row result to beats mapstr")
+		}
+	}
+
+	return nil
+}
+
+// handleRowResult is called for each row result
+func (f *Fetcher) handleRowResult(r *rowsResult) (err error) {
+	// We assign pointers to the destination to pass it to Scan.
+	rawResult := make([]*string, len(r.columnNames))
+	for i := range rawResult {
+		r.dest[i] = &rawResult[i]
+	}
+
+	if err = r.rows.Scan(r.dest...); err != nil {
+		return errors.Wrap(err, "error scanning row of result")
+	}
+
+	// Now we need to get the values of the pointers back into a normal
+	// map[string]interface{} to use it with the schema.
+	mapOfResults := make(map[string]interface{})
+
+	for i, res := range rawResult {
+		if res != nil {
+			mapOfResults[r.columnNames[i]] = *res
+		}
+	}
+
+	result, err := f.schema.Apply(mapOfResults)
+	if err != nil {
+		r.reporter.Error(err)
```

Comment:
Maybe it worths to wrap this error with a message about this failing when applying the schema.

Reply:
Done :)